### PR TITLE
[AppConfiguration] Fix SDK API breaking for contentType

### PR DESCRIPTION
**Pull request by msyyc**
`contentTypeHeader` as method signature is breaking compared with last release version. So we need `@friendlyName` to make the name same as before.
Update:
Some reviewers have concern about `@friendlyName`, so I change the property name directly then use `Body` to warp `KeyValue` to avoid tsp compiler error. As for template `ResourceRead` that even `Body` doesn't work, I take some effort to write  with more basic template `Foundations.ResourceOperation` and make sure generated swagger has no change.

### CODE REVIEW

File: specification/appconfiguration/AppConfiguration/routes.tsp [hunk 1]
```diff
@@ -36,6 +36,7 @@ alias syncTokenHeader = {
 alias contentTypeHeader<T extends string> = {
   @doc("Content-Type header")
   @header("Content-Type")
+  @friendlyName("contentType")
   contentTypeHeader?: T | "application/problem+json";
```
Comment (weidongxu-microsoft) [inline specification/appconfiguration/AppConfiguration/routes.tsp:40]: Why not rename the property directly to `contentType`?

I didn't find `contentTypeHeader` in swagger, so I assume the `contentTypeHeader` does not matter.
https://github.com/Azure/azure-rest-api-specs/blob/main/specification/appconfiguration/data-plane/Microsoft.AppConfiguration/stable/2023-11-01/appconfiguration.json

Comment (lmazuel) [reply specification/appconfiguration/AppConfiguration/routes.tsp:40]: @catalinaperalta , any reasons why we couldn't just name is `contentType` ?

Comment (catalinaperalta) [reply specification/appconfiguration/AppConfiguration/routes.tsp:40]: I dont think there should be a problem renaming this directly. Try renaming it to just contentType and make sure to regenerate the openapi spec since this would change it (run `tsp compile .`) and push the update openapi spec to this PR since you need to keep it in sync with the tsp definitions. Let me know if you run into any issues.

Comment (weidongxu-microsoft) [reply specification/appconfiguration/AppConfiguration/routes.tsp:40]: One possible cause is that there is another `contentType` in body (and that's flattened).

One caution is that the `contentTypeHeader` (after `@friendlyName`) and `contentType` (in body) would have a same name. We may want make sure JS dev is OK with it (I think JS may have named property of headers in response).

Comment (msyyc) [reply specification/appconfiguration/AppConfiguration/routes.tsp:40]: @lmazuel @catalinaperalta Yesterday I tried to rename tsp directly but meet compiler error since there is already a property named "contentType" in "[KeyValue](https://github.com/Azure/azure-rest-api-specs/blob/c65e8cbbb18fc9d79165f12291d38d91eb340f0f/specification/appconfiguration/AppConfiguration/models.tsp#L182)"

image

Then I find the easiest way is to use `@friendlyName`.

Comment (msyyc) [reply specification/appconfiguration/AppConfiguration/routes.tsp:40]: > One possible cause is that there is another `contentType` in (implicit) body.
> 
> One caution is that the `contentTypeHeader` (after `@friendlyName`) and `contentType` (in body) would have a same name. We may want to make sure JS dev is OK with it (I think JS may have named property of headers in response).

Add JS owner @MaryGao @qiaozha for awareness.

Comment (lmazuel) [reply specification/appconfiguration/AppConfiguration/routes.tsp:40]: Talked with @timotheeguerin offline, we can make this works without the friendly with a slight change in the operation return definition:
```typescript
  Body<KeyValue> &
    AppConfigResponseHeaders &
    contentTypeHeader<"application/vnd.microsoft.appconfig.kv+json">
```

See the addition of `Body` around `KeyVault`, which helps the TypeSpec emitter to understand there should be no conflict.

**Review by qiaozha — APPROVED** [review]
LGTM

File: specification/appconfiguration/AppConfiguration/routes.tsp (hunk 1 above)
Comment (msyyc) [reply specification/appconfiguration/AppConfiguration/routes.tsp]: After update with @timotheeguerin's suggestion, only one error left: 
image

It is related with Template `ResourceRead`:

image

I understand that we need to customize this template like:?

```
  @Foundations.Private.ensureVerb("ResourceRead", "GET")
  @readsResource(Resource)
  ResourceRead<
    Resource extends TypeSpec.Reflection.Model,
    Traits extends TypeSpec.Reflection.Model = {}
  > is Foundations.ResourceOperation<
    Resource,
    TraitProperties<Traits & InterfaceTraits, TraitLocation.Parameters, TraitContext.Read>,
    Body<Resource> &
      TraitProperties<Traits & InterfaceTraits, TraitLocation.Response, TraitContext.Read>,
    Traits & InterfaceTraits,
    ErrorResponse
  >;
```

Comment (timotheeguerin) [reply specification/appconfiguration/AppConfiguration/routes.tsp]: Can you show how this template is called currently where it cause this error

File: specification/appconfiguration/AppConfiguration/routes.tsp [hunk 2]
```diff
@@ -9,6 +9,13 @@ using Azure.Core.Traits;
 
 namespace AzureAppConfiguration;
 
+alias Body<T, TName extends valueof string = "body", TDoc extends valueof string = "The body parameter."> = {
+  @doc(TDoc)
+  @friendlyName(TName)
```
Comment (timotheeguerin) [inline specification/appconfiguration/AppConfiguration/routes.tsp:14]: Do not use friendly name like that

File: specification/appconfiguration/AppConfiguration/routes.tsp [hunk 3]
```diff
@@ -9,6 +9,13 @@ using Azure.Core.Traits;
 
 namespace AzureAppConfiguration;
 
+alias Body<T, TName extends valueof string = "body", TDoc extends valueof string = "The body parameter."> = {
```
Comment (timotheeguerin) [inline specification/appconfiguration/AppConfiguration/routes.tsp:12]: There is also already bodyparameter type in azure core

Comment (catalinaperalta) [reply specification/appconfiguration/AppConfiguration/routes.tsp:12]: I saw the Body<> template in typespec/http and tested it out in the playground

File: specification/appconfiguration/AppConfiguration/routes.tsp (hunk 2 above)
Comment (catalinaperalta) [reply specification/appconfiguration/AppConfiguration/routes.tsp:14]: If the issue we're trying to solve is only for client libraries why arent we just using clientName?

File: specification/appconfiguration/AppConfiguration/routes.tsp [hunk 4]
```diff
@@ -36,7 +43,7 @@ alias syncTokenHeader = {
 alias contentTypeHeader<T extends string> = {
   @doc("Content-Type header")
   @header("Content-Type")
-  contentTypeHeader?: T | "application/problem+json";
+  contentType?: T | "application/problem+json";
```
Comment (catalinaperalta) [inline specification/appconfiguration/AppConfiguration/routes.tsp:46]: Could we put a clientName decorator over this?

File: specification/appconfiguration/AppConfiguration/routes.tsp (hunk 1 above)
Comment (catalinaperalta) [reply specification/appconfiguration/AppConfiguration/routes.tsp]: @msyyc we shouldnt use internal decorators in service specs, you can create a custom template if you need, but maybe we can just use the clientName decorator?

Comment (timotheeguerin) [reply specification/appconfiguration/AppConfiguration/routes.tsp]: Yeah should definitely not call any internal decorators those will have no non breaking requirement.

In the case of this template this decorator is just to add some validation so you would just not use it but I don't think a custom template is the right solution here.

**Comment by msyyc** [comment]
PR progress:
- Solution A: use @friendlyName but got comment of https://github.com/Azure/azure-rest-api-specs/pull/29887#discussion_r1682513902 then I switch to solution B
- Solution B: change `contentTypeHeader` directly then get error of https://github.com/Azure/azure-rest-api-specs/pull/29887#discussion_r1685972559 and timothee doesn't advise to write a customized template then I switch to solution C
- Solution C: use @clientName but I tried the following ways and all fails:
image

Now the only solution left is solution A which is also easiest. @catalinaperalta @timotheeguerin

**Review by weidongxu-microsoft — APPROVED** [review]
as long as service is good with this change

File: specification/appconfiguration/AppConfiguration/routes.tsp [hunk 5]
```diff
@@ -259,45 +259,53 @@ op checkKeyValues is appConfigOperation<
...
 >;
 
+#suppress "@azure-tools/typespec-azure-core/use-standard-operations" "Pre-existing API contract"
 @summary("Gets a single key-value.")
 @doc("Gets a single key-value.")
-op getKeyValue is standardOps.ResourceRead<
-  KeyValue,
-  QueryParametersTrait<{
+@get
+op getKeyValue is Foundations.ResourceOperation<
```
Comment (catalinaperalta) [inline specification/appconfiguration/AppConfiguration/routes.tsp:266]: Why cant we use the standard resource read operation anymore?

File: specification/appconfiguration/AppConfiguration/routes.tsp [hunk 6]
```diff
@@ -346,7 +354,7 @@ op putKeyValue is appConfigOperation<
     @body
     entity?: KeyValue;
   },
-  KeyValue &
+  Body<KeyValue> &
```
Comment (catalinaperalta) [inline specification/appconfiguration/AppConfiguration/routes.tsp:357]: Why is this necessary?

Comment (msyyc) [reply specification/appconfiguration/AppConfiguration/routes.tsp:357]: https://github.com/Azure/azure-rest-api-specs/pull/29887#discussion_r1684975812

File: specification/appconfiguration/AppConfiguration/routes.tsp (hunk 5 above)
Comment (msyyc) [reply specification/appconfiguration/AppConfiguration/routes.tsp:266]: Standard template will cause compiler error https://github.com/Azure/azure-rest-api-specs/pull/29887#discussion_r1685972559 so we have to change to another writing way

**Review by catalinaperalta — APPROVED** [review]
